File: Sources/BudgetCore/DebtProjection.swift
import Foundation

public enum DebtPaymentFrequency: String, Sendable { case weekly, biweekly, monthly }
public enum DebtMinimumPaymentRule: String, Sendable { case fixed, percentage, greaterOf }
public enum DebtProjectionStatus: String, Sendable { case paidOff, nonAmortizing, iterationLimit }
public enum DebtPayoffStrategy: String, Sendable { case avalanche, snowball, custom }

public struct DebtProjectionTerms: Sendable {
    public let annualRateBasisPoints: Int64
    public let frequency: DebtPaymentFrequency
    public let scheduledPaymentMinor: Int64?
    public let minimumRule: DebtMinimumPaymentRule?
    public let minimumPaymentMinor: Int64?
    public let minimumRateBasisPoints: Int64?
    public let promotionalRateBasisPoints: Int64?
    public let promotionalEndsOn: Date?

    public init(annualRateBasisPoints: Int64, frequency: DebtPaymentFrequency, scheduledPaymentMinor: Int64? = nil, minimumRule: DebtMinimumPaymentRule? = nil, minimumPaymentMinor: Int64? = nil, minimumRateBasisPoints: Int64? = nil, promotionalRateBasisPoints: Int64? = nil, promotionalEndsOn: Date? = nil) {
        self.annualRateBasisPoints = annualRateBasisPoints; self.frequency = frequency
        self.scheduledPaymentMinor = scheduledPaymentMinor; self.minimumRule = minimumRule
        self.minimumPaymentMinor = minimumPaymentMinor; self.minimumRateBasisPoints = minimumRateBasisPoints
        self.promotionalRateBasisPoints = promotionalRateBasisPoints; self.promotionalEndsOn = promotionalEndsOn
    }
}

public struct DebtProjectionPoint: Equatable, Sendable {
    public let paymentNumber: Int; public let paymentDate: Date
    public let startingPrincipalMinor: Int64; public let interestMinor: Int64
    public let paymentMinor: Int64; public let endingPrincipalMinor: Int64
}

public struct DebtProjectionResult: Sendable {
    public let status: DebtProjectionStatus; public let payoffDate: Date?
    public let projectedInterestMinor: Int64; public let projectedTotalCostMinor: Int64
    public let points: [DebtProjectionPoint]
    public var paymentCount: Int { points.count }
}

public struct DebtStrategyInput: Equatable, Sendable {
    public let debtID: String
    public let principalMinor: Int64
    public let annualRateBasisPoints: Int64
    public let plannedPaymentMinor: Int64
    public let promotionalRateBasisPoints: Int64?
    public let promotionalEndsOn: Date?

    public init(debtID: String, principalMinor: Int64, annualRateBasisPoints: Int64, plannedPaymentMinor: Int64, promotionalRateBasisPoints: Int64? = nil, promotionalEndsOn: Date? = nil) {
        self.debtID = debtID; self.principalMinor = principalMinor
        self.annualRateBasisPoints = annualRateBasisPoints
        self.plannedPaymentMinor = plannedPaymentMinor
        self.promotionalRateBasisPoints = promotionalRateBasisPoints
        self.promotionalEndsOn = promotionalEndsOn
    }
}

public struct DebtStrategyAccountResult: Equatable, Sendable {
    public let debtID: String
    public let payoffDate: Date?
    public let payoffMonth: Int?
    public let projectedInterestMinor: Int64
    public let projectedTotalPaidMinor: Int64
}

public struct DebtStrategyProjectionResult: Sendable {
    public let status: DebtProjectionStatus
    public let strategy: DebtPayoffStrategy
    public let rollover: Bool
    public let payoffOrder: [String]
    public let debtFreeDate: Date?
    public let paymentCount: Int
    public let projectedInterestMinor: Int64
    public let projectedTotalPaidMinor: Int64
    public let projectedTotalCostMinor: Int64
    public let debts: [DebtStrategyAccountResult]
}

public enum DebtProjectionEngine {
    public static let maximumPeriods = 1_200

    /// A fixed monthly scenario budget, normalized from the first scheduled payment.
    /// This is not a claim about an issuer's future minimum-payment rules.
    public static func monthlyStrategyPayment(principalMinor: Int64, firstPaymentOn: Date, terms: DebtProjectionTerms) throws -> Int64 {
        guard principalMinor >= 0, (0...100_000).contains(terms.annualRateBasisPoints),
              (0...100_000).contains(terms.promotionalRateBasisPoints ?? 0),
              (0...10_000).contains(terms.minimumRateBasisPoints ?? 0),
              (terms.scheduledPaymentMinor ?? 0) >= 0, (terms.minimumPaymentMinor ?? 0) >= 0 else { throw ProjectionError.invalidInput }
        let periods: Int64 = terms.frequency == .weekly ? 52 : terms.frequency == .biweekly ? 26 : 12
        let rate = terms.promotionalRateBasisPoints != nil && terms.promotionalEndsOn != nil && firstPaymentOn <= terms.promotionalEndsOn! ? terms.promotionalRateBasisPoints! : terms.annualRateBasisPoints
        let interest = try multipliedAndRounded(principalMinor, by: rate, dividedBy: 10_000 * periods)
        let statement = try checkedAdd(principalMinor, interest)
        let payment = try plannedPayment(statement: statement, terms: terms)
        return try multipliedAndRounded(payment, by: periods, dividedBy: 12)
    }

    private static func plannedPayment(statement: Int64, terms: DebtProjectionTerms) throws -> Int64 {
        if let scheduled = terms.scheduledPaymentMinor { return scheduled }
        let percentage = try multipliedAndRounded(statement, by: terms.minimumRateBasisPoints ?? 0, dividedBy: 10_000)
        switch terms.minimumRule {
        case .fixed: return terms.minimumPaymentMinor ?? 0
        case .percentage: return percentage
        case .greaterOf: return max(terms.minimumPaymentMinor ?? 0, percentage)
        case nil: throw ProjectionError.missingPaymentRule
        }
    }

    public static func project(principalMinor: Int64, firstPaymentOn: Date, terms: DebtProjectionTerms, extraPaymentMinor: Int64 = 0, maximumPeriods: Int = maximumPeriods, calendar: Calendar = Calendar(identifier: .gregorian)) throws -> DebtProjectionResult {
        guard principalMinor >= 0, extraPaymentMinor >= 0, (0...100_000).contains(terms.annualRateBasisPoints), (1...self.maximumPeriods).contains(maximumPeriods) else { throw ProjectionError.invalidInput }
        guard (terms.scheduledPaymentMinor ?? 0) >= 0, (terms.minimumPaymentMinor ?? 0) >= 0,
              (0...10_000).contains(terms.minimumRateBasisPoints ?? 0),
              (0...100_000).contains(terms.promotionalRateBasisPoints ?? 0) else { throw ProjectionError.invalidInput }
        if principalMinor == 0 { return .init(status: .paidOff, payoffDate: firstPaymentOn, projectedInterestMinor: 0, projectedTotalCostMinor: 0, points: []) }
        let periods: Int64 = terms.frequency == .weekly ? 52 : terms.frequency == .biweekly ? 26 : 12
        var principal = principalMinor, date = firstPaymentOn, interestTotal: Int64 = 0, paid: Int64 = 0
        var points: [DebtProjectionPoint] = []
        for number in 1...maximumPeriods {
            let rate = terms.promotionalRateBasisPoints != nil && terms.promotionalEndsOn != nil && date <= terms.promotionalEndsOn! ? terms.promotionalRateBasisPoints! : terms.annualRateBasisPoints
            let interest = try multipliedAndRounded(principal, by: rate, dividedBy: 10_000 * periods)
            let statement = try checkedAdd(principal, interest)
            let base = try plannedPayment(statement: statement, terms: terms)
            let planned = try checkedAdd(base, extraPaymentMinor)
            if planned <= interest { return .init(status: .nonAmortizing, payoffDate: nil, projectedInterestMinor: interestTotal, projectedTotalCostMinor: paid, points: points) }
            let payment = min(planned, statement), ending = statement - payment
            points.append(.init(paymentNumber: number, paymentDate: date, startingPrincipalMinor: principal, interestMinor: interest, paymentMinor: payment, endingPrincipalMinor: ending))
            interestTotal = try checkedAdd(interestTotal, interest); paid = try checkedAdd(paid, payment)
            if ending == 0 { return .init(status: .paidOff, payoffDate: date, projectedInterestMinor: interestTotal, projectedTotalCostMinor: paid, points: points) }
            principal = ending
            let component: Calendar.Component = terms.frequency == .monthly ? .month : .day
            let increment = terms.frequency == .weekly ? 7 : terms.frequency == .biweekly ? 14 : 1
            date = calendar.date(byAdding: component, value: increment, to: date)!
        }
        return .init(status: .iterationLimit, payoffDate: nil, projectedInterestMinor: interestTotal, projectedTotalCostMinor: paid, points: points)
    }

    public static func projectStrategy(
        debts: [DebtStrategyInput],
        firstPaymentOn: Date,
        strategy: DebtPayoffStrategy,
        rollover: Bool,
        extraPaymentMinor: Int64 = 0,
        customOrder: [String] = [],
        maximumPeriods: Int = maximumPeriods,
        calendar: Calendar = Calendar(identifier: .gregorian)
    ) throws -> DebtStrategyProjectionResult {
        guard !debts.isEmpty, extraPaymentMinor >= 0, (1...self.maximumPeriods).contains(maximumPeriods),
              debts.allSatisfy({ !$0.debtID.isEmpty && $0.principalMinor >= 0 && $0.plannedPaymentMinor >= 0 && (0...100_000).contains($0.annualRateBasisPoints) }),
              Set(debts.map(\.debtID)).count == debts.count else { throw ProjectionError.invalidInput }
        let ids = debts.map(\.debtID)
        guard debts.allSatisfy({ (0...100_000).contains($0.promotionalRateBasisPoints ?? 0)
            && (($0.promotionalRateBasisPoints == nil) == ($0.promotionalEndsOn == nil)) }) else { throw ProjectionError.invalidInput }
        if strategy == .custom && (customOrder.count != ids.count || Set(customOrder) != Set(ids)) {
            throw ProjectionError.invalidCustomOrder
        }
        var balances = Dictionary(uniqueKeysWithValues: debts.map { ($0.debtID, $0.principalMinor) })
        func ratesOn(_ date: Date) -> [String: Int64] {
            Dictionary(uniqueKeysWithValues: debts.map { debt in
                let rate = debt.promotionalEndsOn.map { date <= $0 } == true ? debt.promotionalRateBasisPoints! : debt.annualRateBasisPoints
                return (debt.debtID, rate)
            })
        }
        var rates = ratesOn(firstPaymentOn)
        let payments = Dictionary(uniqueKeysWithValues: debts.map { ($0.debtID, $0.plannedPaymentMinor) })
        var interestByID = Dictionary(uniqueKeysWithValues: ids.map { ($0, Int64(0)) })
        var paidByID = Dictionary(uniqueKeysWithValues: ids.map { ($0, Int64(0)) })
        var payoffDates: [String: Date] = [:], payoffMonths: [String: Int] = [:]
        var payoffOrder: [String] = [], paymentDate = firstPaymentOn
        var rolloverPool: Int64 = 0

        func priority(_ active: [String]) -> [String] {
            // Non-custom strategies ignore customOrder, including duplicate values.
            let positions = strategy == .custom ? Dictionary(uniqueKeysWithValues: customOrder.enumerated().map { ($0.element, $0.offset) }) : [:]
            return active.sorted { lhs, rhs in
                switch strategy {
                case .avalanche:
                    if rates[lhs]! != rates[rhs]! { return rates[lhs]! > rates[rhs]! }
                    if balances[lhs]! != balances[rhs]! { return balances[lhs]! < balances[rhs]! }
                case .snowball:
                    if balances[lhs]! != balances[rhs]! { return balances[lhs]! < balances[rhs]! }
                    if rates[lhs]! != rates[rhs]! { return rates[lhs]! > rates[rhs]! }
                case .custom:
                    return positions[lhs]! < positions[rhs]!
                }
                return lhs < rhs
            }
        }
        func accountResults() -> [DebtStrategyAccountResult] {
            ids.map { .init(debtID: $0, payoffDate: payoffDates[$0], payoffMonth: payoffMonths[$0], projectedInterestMinor: interestByID[$0]!, projectedTotalPaidMinor: paidByID[$0]!) }
        }
        func result(status: DebtProjectionStatus, date: Date?, count: Int) throws -> DebtStrategyProjectionResult {
            let interest = try interestByID.values.reduce(0, checkedAdd), paid = try paidByID.values.reduce(0, checkedAdd)
            return .init(status: status, strategy: strategy, rollover: rollover, payoffOrder: payoffOrder, debtFreeDate: date, paymentCount: count, projectedInterestMinor: interest, projectedTotalPaidMinor: paid, projectedTotalCostMinor: paid, debts: accountResults())
        }

        for id in priority(ids.filter { balances[$0] == 0 }) {
            payoffOrder.append(id); payoffDates[id] = firstPaymentOn; payoffMonths[id] = 0
        }
        for number in 1...maximumPeriods {
            rates = ratesOn(paymentDate)
            let active = ids.filter { balances[$0]! > 0 }
            if active.isEmpty { return try result(status: .paidOff, date: paymentDate, count: number - 1) }
            let startingTotal = try active.reduce(Int64(0)) { try checkedAdd($0, balances[$1]!) }
            var remaining: [String: Int64] = [:]
            for id in active {
                let interest = try multipliedAndRounded(balances[id]!, by: rates[id]!, dividedBy: 120_000)
                interestByID[id] = try checkedAdd(interestByID[id]!, interest)
                remaining[id] = try checkedAdd(balances[id]!, interest)
            }
            for id in active {
                let payment = min(payments[id]!, remaining[id]!)
                remaining[id]! -= payment; paidByID[id] = try checkedAdd(paidByID[id]!, payment)
            }
            var strategyMoney = try checkedAdd(extraPaymentMinor, rollover ? rolloverPool : 0)
            for id in priority(active.filter { remaining[$0]! > 0 }) {
                let payment = min(strategyMoney, remaining[id]!)
                remaining[id]! -= payment; paidByID[id] = try checkedAdd(paidByID[id]!, payment); strategyMoney -= payment
                if strategyMoney == 0 { break }
            }
            var newlyPaid: [String] = []
            for id in active {
                balances[id] = remaining[id]!
                if balances[id] == 0 {
                    newlyPaid.append(id); payoffDates[id] = paymentDate; payoffMonths[id] = number
                }
            }
            for id in priority(newlyPaid) {
                payoffOrder.append(id)
                if rollover { rolloverPool = try checkedAdd(rolloverPool, payments[id]!) }
            }
            if balances.values.allSatisfy({ $0 == 0 }) { return try result(status: .paidOff, date: paymentDate, count: number) }
            let upcomingRateChange = debts.contains { debt in
                balances[debt.debtID]! > 0 && debt.promotionalEndsOn.map { paymentDate <= $0 } == true
            }
            if try balances.values.reduce(0, checkedAdd) >= startingTotal && newlyPaid.isEmpty && !upcomingRateChange {
                return try result(status: .nonAmortizing, date: nil, count: number)
            }
            paymentDate = calendar.date(byAdding: .month, value: 1, to: paymentDate)!
        }
        return try result(status: .iterationLimit, date: nil, count: maximumPeriods)
    }

    private static func multipliedAndRounded(_ value: Int64, by multiplier: Int64, dividedBy denominator: Int64) throws -> Int64 {
        let quotient = value / denominator, remainder = value % denominator
        let (whole, overflow) = quotient.multipliedReportingOverflow(by: multiplier)
        guard !overflow else { throw ProjectionError.amountOutOfRange }
        // Validated rates and denominators bound the remainder product below Int64.max.
        return try checkedAdd(whole, (remainder * multiplier + denominator / 2) / denominator)
    }
    private static func checkedAdd(_ lhs: Int64, _ rhs: Int64) throws -> Int64 {
        let (value, overflow) = lhs.addingReportingOverflow(rhs)
        guard !overflow else { throw ProjectionError.amountOutOfRange }
        return value
    }
    public enum ProjectionError: LocalizedError {
        case invalidInput, missingPaymentRule, invalidCustomOrder, amountOutOfRange
        public var errorDescription: String? {
            switch self {
            case .amountOutOfRange: "Projection amounts exceed the supported exact-money range. Reduce the scenario amounts."
            case .invalidInput: "Review the projection amounts and rates."
            case .missingPaymentRule: "Add a payment rule in Debt Terms."
            case .invalidCustomOrder: "Include each selected debt once in the custom order."
            }
        }
    }
}

File: server/app/debt_projection.py
"""Deterministic debt payoff projection using integer minor units only.

APR is stored in basis points.  Each payment period accrues simple periodic interest as
``principal_minor * annual_rate_basis_points / (10_000 * periods_per_year)`` and rounds
half up to one minor unit before the payment is applied.  Payments occur at the end of
each period; the final payment is capped to principal plus that period's interest.

This is a scenario model, not an issuer statement reconstruction: it deliberately does
not guess daily-balance, grace-period, fee, or issuer-specific minimum-payment rules.
"""

from __future__ import annotations

from calendar import monthrange
from dataclasses import dataclass
from datetime import date, timedelta
from typing import Literal


Frequency = Literal["weekly", "biweekly", "monthly"]
PaymentRule = Literal["fixed", "percentage", "greater_of"]
Strategy = Literal["avalanche", "snowball", "custom"]
MAX_MONEY = (1 << 63) - 1


def _money(value: int) -> int:
    if not 0 <= value <= MAX_MONEY:
        raise ValueError("Projection amounts exceed the supported exact-money range")
    return value


@dataclass(frozen=True)
class ProjectionTerms:
    annual_rate_basis_points: int
    payment_frequency: Frequency
    scheduled_payment_minor: int | None = None
    minimum_payment_rule: PaymentRule | None = None
    minimum_payment_minor: int | None = None
    minimum_payment_rate_basis_points: int | None = None
    promotional_rate_basis_points: int | None = None
    promotional_ends_on: date | None = None


@dataclass(frozen=True)
class ProjectionPoint:
    payment_number: int
    payment_date: date
    starting_principal_minor: int
    interest_minor: int
    payment_minor: int
    ending_principal_minor: int


@dataclass(frozen=True)
class ProjectionResult:
    status: Literal["paid_off", "non_amortizing", "iteration_limit"]
    payoff_date: date | None
    payment_count: int
    projected_interest_minor: int
    projected_total_cost_minor: int
    points: tuple[ProjectionPoint, ...]

    def __post_init__(self):
        _money(self.projected_interest_minor)
        _money(self.projected_total_cost_minor)


@dataclass(frozen=True)
class StrategyDebt:
    """One debt in a monthly household payoff scenario.

    ``planned_payment_minor`` is the amount already available for this debt each
    month. Strategy scenarios never write this assumption back to the budget.
    """

    debt_id: str
    principal_minor: int
    annual_rate_basis_points: int
    planned_payment_minor: int
    promotional_rate_basis_points: int | None = None
    promotional_ends_on: date | None = None


@dataclass(frozen=True)
class StrategyDebtResult:
    debt_id: str
    payoff_date: date | None
    payoff_month: int | None
    projected_interest_minor: int
    projected_total_paid_minor: int


@dataclass(frozen=True)
class StrategyProjectionResult:
    status: Literal["paid_off", "non_amortizing", "iteration_limit"]
    strategy: Strategy
    rollover: bool
    payoff_order: tuple[str, ...]
    debt_free_date: date | None
    payment_count: int
    projected_interest_minor: int
    projected_total_paid_minor: int
    projected_total_cost_minor: int
    debts: tuple[StrategyDebtResult, ...]

    def __post_init__(self):
        _money(self.projected_interest_minor)
        _money(self.projected_total_paid_minor)


MAX_PROJECTION_PERIODS = 1_200
_PERIODS_PER_YEAR: dict[Frequency, int] = {"weekly": 52, "biweekly": 26, "monthly": 12}


def _round_ratio_half_up(numerator: int, denominator: int) -> int:
    if numerator < 0 or denominator <= 0:
        raise ValueError("projection ratios require non-negative values")
    return (numerator + denominator // 2) // denominator


def _advance(value: date, frequency: Frequency) -> date:
    if frequency == "weekly":
        return value + timedelta(days=7)
    if frequency == "biweekly":
        return value + timedelta(days=14)
    month_index = value.year * 12 + value.month
    year, zero_based_month = divmod(month_index, 12)
    month = zero_based_month + 1
    return date(year, month, min(value.day, monthrange(year, month)[1]))


def _planned_payment(terms: ProjectionTerms, statement_minor: int) -> int:
    if terms.scheduled_payment_minor is not None:
        return terms.scheduled_payment_minor
    percentage = _round_ratio_half_up(
        statement_minor * (terms.minimum_payment_rate_basis_points or 0), 10_000
    )
    if terms.minimum_payment_rule == "fixed":
        return terms.minimum_payment_minor or 0
    if terms.minimum_payment_rule == "percentage":
        return percentage
    if terms.minimum_payment_rule == "greater_of":
        return max(terms.minimum_payment_minor or 0, percentage)
    raise ValueError("terms do not define a payment rule")


def monthly_strategy_payment(
    terms: ProjectionTerms, principal_minor: int, first_payment_on: date
) -> int:
    """Normalize an existing payment rule to an explicit monthly scenario budget."""
    periods = _PERIODS_PER_YEAR[terms.payment_frequency]
    rate = terms.annual_rate_basis_points
    if (
        terms.promotional_rate_basis_points is not None
        and terms.promotional_ends_on is not None
        and first_payment_on <= terms.promotional_ends_on
    ):
        rate = terms.promotional_rate_basis_points
    interest = _round_ratio_half_up(principal_minor * rate, 10_000 * periods)
    periodic = _planned_payment(terms, principal_minor + interest)
    return _round_ratio_half_up(periodic * periods, 12)


def project_debt(
    principal_minor: int,
    first_payment_on: date,
    terms: ProjectionTerms,
    *,
    extra_payment_minor: int = 0,
    max_periods: int = MAX_PROJECTION_PERIODS,
) -> ProjectionResult:
    _money(principal_minor)
    _money(extra_payment_minor)
    _money(terms.scheduled_payment_minor or 0)
    _money(terms.minimum_payment_minor or 0)
    if not 0 <= (terms.minimum_payment_rate_basis_points or 0) <= 10_000 or not 0 <= (terms.promotional_rate_basis_points or 0) <= 100_000:
        raise ValueError("invalid projection rate")
    if principal_minor < 0 or extra_payment_minor < 0 or not 0 <= terms.annual_rate_basis_points <= 100_000:
        raise ValueError("invalid projection input")
    if max_periods < 1 or max_periods > MAX_PROJECTION_PERIODS:
        raise ValueError("invalid projection bound")
    if principal_minor == 0:
        return ProjectionResult("paid_off", first_payment_on, 0, 0, 0, ())

    principal = principal_minor
    payment_date = first_payment_on
    total_interest = 0
    total_paid = 0
    points: list[ProjectionPoint] = []
    periods = _PERIODS_PER_YEAR[terms.payment_frequency]

    for number in range(1, max_periods + 1):
        rate = terms.annual_rate_basis_points
        if terms.promotional_rate_basis_points is not None and terms.promotional_ends_on is not None and payment_date <= terms.promotional_ends_on:
            rate = terms.promotional_rate_basis_points
        interest = _round_ratio_half_up(principal * rate, 10_000 * periods)
        statement = _money(principal + interest)
        planned = _money(_planned_payment(terms, statement) + extra_payment_minor)
        if planned <= interest:
            return ProjectionResult("non_amortizing", None, number - 1, total_interest, total_paid, tuple(points))
        payment = min(planned, statement)
        ending = statement - payment
        points.append(ProjectionPoint(number, payment_date, principal, interest, payment, ending))
        total_interest += interest
        total_paid += payment
        if ending == 0:
            return ProjectionResult("paid_off", payment_date, number, total_interest, total_paid, tuple(points))
        principal = ending
        payment_date = _advance(payment_date, terms.payment_frequency)

    return ProjectionResult("iteration_limit", None, max_periods, total_interest, total_paid, tuple(points))


def project_debt_strategy(
    debts: list[StrategyDebt] | tuple[StrategyDebt, ...],
    first_payment_on: date,
    *,
    strategy: Strategy,
    rollover: bool,
    extra_payment_minor: int = 0,
    custom_order: list[str] | tuple[str, ...] = (),
    max_periods: int = MAX_PROJECTION_PERIODS,
) -> StrategyProjectionResult:
    """Project a monthly multi-debt strategy without mutating financial truth.

    Every active debt first receives its existing planned payment. The scenario's
    extra amount is then applied in the selected order. With rollover enabled,
    the planned payment of a paid debt becomes additional strategy money beginning
    the following month. Any unused amount from a final capped payment cascades in
    the same month so scenario money is not silently discarded.
    """

    if strategy not in {"avalanche", "snowball", "custom"} or not debts or extra_payment_minor < 0 or not 1 <= max_periods <= MAX_PROJECTION_PERIODS:
        raise ValueError("invalid strategy projection input")
    _money(extra_payment_minor)
    for debt in debts:
        _money(debt.principal_minor)
        _money(debt.planned_payment_minor)
    ids = [item.debt_id for item in debts]
    if any(not item.debt_id or item.principal_minor < 0 or item.planned_payment_minor < 0 or not 0 <= item.annual_rate_basis_points <= 100_000 for item in debts):
        raise ValueError("invalid strategy debt")
    if any(not 0 <= (item.promotional_rate_basis_points or 0) <= 100_000
           or (item.promotional_rate_basis_points is None) != (item.promotional_ends_on is None) for item in debts):
        raise ValueError("invalid promotional terms")
    if len(set(ids)) != len(ids):
        raise ValueError("strategy debt IDs must be unique")
    custom = tuple(custom_order)
    if strategy == "custom" and (len(custom) != len(ids) or set(custom) != set(ids)):
        raise ValueError("custom order must contain every debt exactly once")

    balances = {item.debt_id: item.principal_minor for item in debts}
    def rates_on(payment_date: date) -> dict[str, int]:
        return {item.debt_id: item.promotional_rate_basis_points
                if item.promotional_ends_on is not None and payment_date <= item.promotional_ends_on
                else item.annual_rate_basis_points for item in debts}
    rates = rates_on(first_payment_on)
    payments = {item.debt_id: item.planned_payment_minor for item in debts}
    interest_by_id = {item.debt_id: 0 for item in debts}
    paid_by_id = {item.debt_id: 0 for item in debts}
    payoff_dates: dict[str, date] = {}
    payoff_months: dict[str, int] = {}
    payoff_order: list[str] = []
    payment_date = first_payment_on
    rollover_pool = 0

    def priority(active: list[str]) -> list[str]:
        if strategy == "avalanche":
            return sorted(active, key=lambda item: (-rates[item], balances[item], item))
        if strategy == "snowball":
            return sorted(active, key=lambda item: (balances[item], -rates[item], item))
        position = {item: index for index, item in enumerate(custom)}
        return sorted(active, key=position.__getitem__)

    initially_paid = [item for item in ids if balances[item] == 0]
    for item in priority(initially_paid):
        payoff_order.append(item)
        payoff_dates[item] = first_payment_on
        payoff_months[item] = 0

    if len(initially_paid) == len(ids):
        results = tuple(StrategyDebtResult(item, first_payment_on, 0, 0, 0) for item in ids)
        return StrategyProjectionResult(
            "paid_off", strategy, rollover, tuple(payoff_order), first_payment_on, 0, 0, 0, 0, results
        )

    for number in range(1, max_periods + 1):
        rates = rates_on(payment_date)
        active = [item for item in ids if balances[item] > 0]
        if not active:
            break
        starting_total = _money(sum(balances[item] for item in active))
        statements: dict[str, int] = {}
        for item in active:
            interest = _round_ratio_half_up(balances[item] * rates[item], 10_000 * 12)
            interest_by_id[item] += interest
            statements[item] = _money(balances[item] + interest)

        # Existing planned payments remain assigned to their debts. Rollover is
        # added only after a debt is gone, never while it still has a balance.
        remaining = dict(statements)
        for item in active:
            payment = min(payments[item], remaining[item])
            remaining[item] -= payment
            paid_by_id[item] += payment

        strategy_money = _money(extra_payment_minor + (rollover_pool if rollover else 0))
        for item in priority([value for value in active if remaining[value] > 0]):
            payment = min(strategy_money, remaining[item])
            remaining[item] -= payment
            paid_by_id[item] += payment
            strategy_money -= payment
            if strategy_money == 0:
                break

        newly_paid: list[str] = []
        for item in active:
            balances[item] = remaining[item]
            if balances[item] == 0:
                newly_paid.append(item)
                payoff_dates[item] = payment_date
                payoff_months[item] = number
        for item in priority(newly_paid):
            payoff_order.append(item)
            if rollover:
                rollover_pool = _money(rollover_pool + payments[item])

        if all(value == 0 for value in balances.values()):
            results = tuple(
                StrategyDebtResult(item, payoff_dates.get(item), payoff_months.get(item), interest_by_id[item], paid_by_id[item])
                for item in ids
            )
            total_interest = sum(interest_by_id.values())
            total_paid = sum(paid_by_id.values())
            return StrategyProjectionResult("paid_off", strategy, rollover, tuple(payoff_order), payment_date, number, total_interest, total_paid, total_paid, results)

        # If aggregate principal did not decline, the fixed household payment
        # budget cannot amortize this scenario. Return an explicit typed result
        # instead of manufacturing a debt-free date.
        upcoming_rate_change = any(balances[item.debt_id] > 0 and item.promotional_ends_on is not None
                                   and payment_date <= item.promotional_ends_on for item in debts)
        if _money(sum(balances.values())) >= starting_total and not newly_paid and not upcoming_rate_change:
            results = tuple(
                StrategyDebtResult(item, payoff_dates.get(item), payoff_months.get(item), interest_by_id[item], paid_by_id[item])
                for item in ids
            )
            total_interest = sum(interest_by_id.values())
            total_paid = sum(paid_by_id.values())
            return StrategyProjectionResult("non_amortizing", strategy, rollover, tuple(payoff_order), None, number, total_interest, total_paid, total_paid, results)
        payment_date = _advance(payment_date, "monthly")

    results = tuple(
        StrategyDebtResult(item, payoff_dates.get(item), payoff_months.get(item), interest_by_id[item], paid_by_id[item])
        for item in ids
    )
    total_interest = sum(interest_by_id.values())
    total_paid = sum(paid_by_id.values())
    status: Literal["paid_off", "non_amortizing", "iteration_limit"] = "iteration_limit"
    if total_paid == 0 or all(payments[item] == 0 for item in ids) and extra_payment_minor == 0:
        status = "non_amortizing"
    return StrategyProjectionResult(status, strategy, rollover, tuple(payoff_order), None, max_periods, total_interest, total_paid, total_paid, results)
